maxprocs/maxprocs.go

```diff
@@ -103,17 +114,92 @@ func SetWithConfig(c *config.Config, opts ...Option) {
 		minProcs:              c.GetInt("MinProcs", defaultMinProcs),
 		cpuRequestsMultiplier: c.GetFloat64("RequestsMultiplier", defaultCPURequestsMultiplier),
 		roundQuotaFunc:        roundQuotaCeil,
+		stop:                  make(chan os.Signal, 1),
 	}
 	for _, opt := range opts {
 		opt(conf)
 	}
 
-	Set(c.GetString("Requests", "1"),
+	var (
+		fileMode     bool
+		requests     = c.GetString("Requests", "1")
+		requestsFile = c.GetString("RequestsFile", "/etc/podinfo/cpu_requests")
+	)
+	if data, err := os.ReadFile(requestsFile); err == nil && len(data) > 0 {
+		fileMode = true
+		requests = strings.TrimSpace(string(data)) + "m"
+		conf.logger.Infon("Using CPU requests from file",
+			logger.NewStringField("requests", requests),
+			logger.NewStringField("file", requestsFile),
+		)
+	}
+
+	Set(requests,
 		WithLogger(conf.logger),
 		WithMinProcs(conf.minProcs),
 		WithCPURequestsMultiplier(conf.cpuRequestsMultiplier),
 		WithRoundQuotaFunc(conf.roundQuotaFunc),
 	)
+
+	if fileMode && c.GetBool("Watch", true) {
+		conf.logger.Infon("Starting file watcher to monitor CPU requests changes",
+			logger.NewStringField("file", requestsFile),
+		)
+		signal.Notify(conf.stop, os.Interrupt, syscall.SIGTERM)
+		go watchFile(conf, requestsFile)
```

Comment:
Instead of spawning a goroutine that the caller cannot control or know when it finishes or fails, I would prefer us to have a new exported function instead, e.g.
```go
// WatchAndSetWithConfig watches the specified file for changes in cpu requests value and updates GOMAXPROCS accordingly.
// The function will return an error if it fails to set up the file watcher. After context cancellation, it stops watching the file and returns nil.
func WatchAndSetWithConfig(ctx context.Context, requestsFile string, c *config.Config, opts ...Option) error {
```

Reply:
@atzoum my goal is to keep using the library like we do in the other services, i.e. with minimal configuration:

```go
import (
     _ "github.com/rudderlabs/rudder-go-kit/maxprocs"
)
```

That's all you'll need with my approach. The idea is to just use env vars and not use any Go code at all.

> Instead of spawning a goroutine that the caller cannot control [...]

You can set `Watch` to `false` if you don't want to start it and you can pass a channel if you want to stop it. 

What else would you like to control?

Reply:
if we only expect this package to be used by just importing it, then we don't need any exported function/variable contained in it.

Reply:
Good idea, I can make everything unexported :+1: